src/libraries/Common/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Pkcs12.cs

```diff
@@ -370,16 +366,56 @@ private static void ProcessSafeContents(
                                     // so always preserve it.
                                     Oids.MsPkcs12MachineKeySet => true,
                                     Oids.Pkcs9FriendlyName => limits.PreserveKeyName,
-                                    Oids.MsPkcs12KeyProviderName => limits.PreserveStorageProvider,
+                                    Oids.MsPkcs12KeyProviderName => ProviderNameIsRelevant,
                                     _ => limits.PreserveUnknownAttributes,
                                 });
                     }
 
-                    if (!loaderLimits.PreserveStorageProvider && s_syntheticKspAttribute.HasValue)
+                    AttributeAsn? providerName = null;
+
+                    if (ProviderNameIsRelevant)
                     {
-                        int newCount = (bag.BagAttributes?.Length).GetValueOrDefault(0) + 1;
-                        Array.Resize(ref bag.BagAttributes, newCount);
-                        bag.BagAttributes[newCount - 1] = s_syntheticKspAttribute.GetValueOrDefault();
+                        if (!loaderLimits.PreserveStorageProvider)
+                        {
+                            providerName = DetermineStorageProvider(
+                                bag.BagAttributes,
+                                bagState.StorageFlags,
+                                ref machineKeyState);
+                        }
+
+                        // If DetermineStorageProvider returns null, leave an existing provider alone,
+                        // but if there's no provider, add the synthetic KSP attribute.
+
+                        // Otherwise, replace any existing provider with what it returned,
+                        // or add what it said if there are no provider attributes.
+
+                        bool hadAttribute = false;
```

Comment:
Past-tense here seems weird.

Reply:
I don't know "had" to implies it... went somewhere. Like it's not there anymore.

I don't have strong opinions, but I paused when I saw it. "Are we removing it or something?"